stan/math/rev/mat/functor/cvodes_integrator.hpp

```diff
@@ -158,26 +133,17 @@ class cvodes_integrator {
       // for the stiff solvers we need to reserve additional
       // memory and provide a Jacobian function call
       // new API since 3.0.0: create matrix object and linear solver object
-      SUNMatrix A{SUNDenseMatrix(N, N)};
-      SUNLinearSolver LS{SUNDenseLinearSolver(cvodes_state, A)};
-      cvodes_check_flag(CVDlsSetLinearSolver(cvodes_mem, LS, A),
-                        "CVDlsSetLinearSolver");
+      cvodes_check_flag(
```

Comment:
@wds15 or @yizhang-cae, can you take a look at these few lines and compare to `develop`? I'm not an expert in how CVODES should be used, but this looks like it's not using CVODES 3.0.

Reply:
I have moved the resource allocation things into `cvodes_ode_data`. This ensures proper resource deallocation whenever we throw an exception. I think that the current version in develop leaks memory whenever an exception is fired from the ODE RHS since those additional matrices are not taken off the memory. This is now ensured by using RAII and relying essentially on the destructor of `cvodes_ode_data` which will be called even if an exception is fired.